### PR TITLE
DrivAerML: Weight decay ablation (WD=0/5e-3/2e-2)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

WD=1e-2 was set in the golden config but has never been ablated on DrivAerML. With only 394 training batches per epoch on 3D surface data, strong weight decay (1e-2) may be reducing the model's effective capacity — acting like implicit capacity reduction on top of the already-limited data regime. WD=0 removes this constraint entirely; WD=5e-3 (half baseline) gives lighter regularization; WD=2e-2 (double baseline) tests whether more regularization helps with the small dataset. The interaction with the cosine LR schedule also makes WD worth probing — at T_max=30, the LR drops low enough that WD could dominate.

## Baseline
val_primary/surface_rel_l2_pct = **4.619%** (PR #2691, 4L/512d/8H, T_max=30, lr=5e-4, AdamW, WD=1e-2)

## Runs

All runs keep golden config values: 4L/512d/8H, lr=5e-4, gc=1.0, T_max=30, AdamW, no-use-ema, enable-fourier, 50k surface points.

**Run 1 (CUDA 0): WD=0**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-WD --wandb-name drivaerml-WD0
```

**Run 2 (CUDA 1): WD=5e-3**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 5e-3 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-WD --wandb-name drivaerml-WD5e3
```

**Run 3 (CUDA 2): WD=2e-2**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 2e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-WD --wandb-name drivaerml-WD2e2
```

## Expected Outcome
val_primary/surface_rel_l2_pct < **4.619%**

Report val_primary/surface_rel_l2_pct for each run. Note whether training was stable and whether any run showed signs of underfitting (WD=0) or overfitting (WD=2e-2).

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shinobu:

## Results

### Summary

| Run | Weight Decay | Best Epoch | val_primary/surface_rel_l2_pct | vs Baseline (4.619%) | Stable? |
|-----|-------------|------------|-------------------------------|---------------------|---------|
| **Baseline** | **1e-2** | **256** | **4.619%** | — | **Yes** |
| WD=0 | 0 | 163 | 6.218% | +34.6% worse | **DIVERGED at ep198** |
| WD=5e-3 | 5e-3 | 220 | 4.986% | +7.9% worse | Yes |
| WD=2e-2 | 2e-2 | 220 | 5.223% | +13.1% worse | Yes |

**None of the ablation runs beat the baseline.** WD=1e-2 remains the optimal weight decay for DrivAerML.

### Detailed Results

**Run 1: WD=0 (CUDA 0)** — W&B run: `faleubiu`
- Best: **6.218%** at epoch 163/243 (loss=0.0103, grad_norm_mean=0.354, only 3 clip events)
- **Catastrophic divergence at epoch 198**: gradient norms → infinity, model collapsed to 61.84%
- Final (ep243): val=61.84%, loss=0.666 (completely broken)
- Without any weight decay, the model's weights grew unconstrained. The low clip events at best epoch (3 vs 350+ in early training) show the model briefly found a very flat basin, but without WD to constrain weight magnitudes, it eventually escaped and diverged catastrophically.

**Run 2: WD=5e-3 (CUDA 1)** — W&B run: `8itei37u`
- Best: **4.986%** at epoch 220/245 (loss=0.0061, grad_norm_mean=0.166, 0 clip events)
- Final (ep245): val=5.096% (mild overshoot, stable)
- Training remained stable throughout. Halving WD from 1e-2 to 5e-3 produced slightly less regularization — the model found a reasonable basin but 7.9% shallower than baseline. The lower grad norms (0.166 vs baseline's typical 0.2-0.3) suggest the lighter decay allows more freedom but doesn't reach as deep a minimum.

**Run 3: WD=2e-2 (CUDA 2)** — W&B run: `2amty67l`
- Best: **5.223%** at epoch 220/247 (loss=0.0068, grad_norm_mean=0.194, 0 clip events)
- Final (ep247): val=7.413% (noticeable degradation from best)
- Doubling WD to 2e-2 over-regularized: effective capacity was too constrained, preventing the model from reaching the deeper basins found at WD=1e-2. The final-epoch degradation (5.22% → 7.41%) suggests the stronger decay makes the loss landscape more oscillatory at this budget.

### Commands Used

```bash
# WD=0
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent shinobu --wandb-group drivaerml-WD --wandb-name "shinobu/drivaerml-WD0"

# WD=5e-3
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 5e-3 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent shinobu --wandb-group drivaerml-WD --wandb-name "shinobu/drivaerml-WD5e3"

# WD=2e-2
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 2e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent shinobu --wandb-group drivaerml-WD --wandb-name "shinobu/drivaerml-WD2e2"
```

### What happened

**WD=1e-2 is the sweet spot for DrivAerML.** The ablation confirms that the golden config's weight decay was well-chosen:

1. **WD=0 is unstable**: Without weight decay, the model trains well initially but catastrophically diverges at ~ep198. Weight magnitudes grow unchecked, and even with grad-clip=1.0, the model eventually escapes stability. This is the strongest evidence that WD is load-bearing on DrivAerML.

2. **WD=5e-3 under-regularizes slightly**: The closest result to baseline (4.99% vs 4.62%) but doesn't quite reach the same depth. With only 394 training batches/epoch, the model needs the full 1e-2 decay to properly constrain the optimization landscape.

3. **WD=2e-2 over-regularizes**: At double the baseline decay, the model loses effective capacity. The 5.22% best is worse than both baseline and WD=5e-3, confirming that DrivAerML needs the model's full capacity given its limited training data.

The WD=0 divergence at ep198 is particularly interesting — it confirms that weight decay isn't just a nice-to-have but is essential for training stability on DrivAerML's 3D surface data. The grad-clip alone cannot prevent weight explosion.

### Suggested follow-ups

- **WD=7.5e-3**: The gap between WD=5e-3 (4.99%) and WD=1e-2 (4.62%) suggests the optimum might be closer to 1e-2 but could be explored more finely.
- **WD schedule**: Start with WD=5e-3 for the first half (more capacity exploration) then increase to WD=1e-2 for stability. This could combine the early exploration freedom with late training stability.
- **WD + gc interaction**: Test WD=1e-2 with gc=0.5 — the AirfRANS results showed gc=0.5 can find deeper basins. The WD ablation confirms WD=1e-2 is stable, so reducing gc is the next lever to probe.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2851 because ADVISOR: CLOSE — WD ablation conclusive. WD=0 (6.22%, diverged), WD=5e-3 (4.99%), WD=2e-2 (5.22%). WD=1e-2 confirmed optimal. WD+gc compound being tested in new wave..
